### DISCUSSION TITLE
Adding more tests and increasing the code coverage

### DISCUSSION THREAD

**Discussion by Sonichigo** [zitadel/zitadel#6946] [Contributors]
I was going through the codebase and pr, to see how ways to contribute to the  zitadel. I noticed that there is a folder for e2e tests written but couldn't see more integrations, as well as i noticed that the code coverage is around 51%. I use a tool called Keploy to for testing and generating data mocks, I think it would be a nice to have some more tests and a bit higher codecoverage. Can i contirbute in this way?

**Comment by hifabienne** [comment]
hei @Sonichigo 
We have added a new api recently where we now add tests for each api request. as we will migrate all the requests to the new apis in the future, we will hopefully get the testcoverage up.

@eliobischof @adlerhurst can you tell more about it and help with the question?

**Reply by Sonichigo** [reply]
hey @hifabienne ! that sounds intriguing! could you please provide more about it ? @eliobischof @adlerhurst

**Reply by Sonichigo** [reply]
@fforootd @muhlemmer can you help me out in understanding ?

**Reply by muhlemmer** [reply]
As we use gRPC for our APIs, we write integration tests as a Go gRPC client. The tests are directly alongside the server package they are testing.

See the contribution guidelines and Makefile  on how integration tests are run.

**Reply by Sonichigo** [reply]
@muhlemmer We now have support for gRPC Client, can i make the PoC PR and you can lmk if that looks promising ?

**Reply by muhlemmer** [reply]
As I mentioned we already expanded test coverage with integrations tests using the generated gRPC client. For example here: 

https://github.com/zitadel/zitadel/blob/main/internal/api/grpc/user/v2/passkey_integration_test.go

As you can see our tests cover complex subjects, not just testing input and output on an API. For example, in the above file we test adding passkeys (webAuthN) to a user. In order to do that we need to mock the browser for signing the response data. I'm not sure if Keploy would be up for the task. IDK, maybe on "easy" endpoints it would work, but it is the complex ones that are the most interesting to test.

Just for full disclosure, because You talk about "_We have..._" and I see you are part of the keyploy organization:

1. Are you a zitadel user and are you benefiting from the enhancement yourself?
2. Are you looking to contribute just promote your tool?

I'm mainly concerned about receiving a contribution and our developers needing to learn to the new tool to maintain it. (Technical debt).
